src/main/java/org/terasology/scenario/internal/ui/LogicTreeMenuTreeBuilder.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2017 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.scenario.internal.ui;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.rendering.nui.NUIManager;
+import org.terasology.rendering.nui.contextMenu.MenuTree;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class LogicTreeMenuTreeBuilder {

```

Comment:
I am also someone who likes to use the builder pattern. However this class is a bit confusing to me. It knows actions by string and also uses them to construct something but does get the logic from outside.

What I don't like about it is the complexity. As anothe developer i have to basically read the complete class in order to understand your other code. Also it is not really a clean separation of concerns/roles. 

e.g. a typical use case for a builder like this could be the separation of the technical construction of something (done by the builder) and the logic of what should be put together (the caller of the builder). In your case the createPrimaryContextMenu contains logic that decised what should be put together which is normally not within the builder. Also it is very strange that a builder has to get data from outside that it then uses to build something or not. 

So in your case I would first have gone for a solution without the builder pattern and then see later on if it makes sense to introduce one.